http3/server.go

```diff
@@ -497,19 +538,51 @@ func (s *Server) handleConn(serverCtx context.Context, conn quic.Connection) err
 	// Process all requests immediately.
 	// It's the client's responsibility to decide which requests are eligible for 0-RTT.
 	for {
-		str, datagrams, err := hconn.acceptStream(serverCtx)
+		str, datagrams, err := hconn.acceptStream(s.graceCtx)
 		if err != nil {
-			// close the connection if the server was closed
-			if errors.Is(err, context.Canceled) {
+			// not gracefully closed, close the connection immediately
+			if s.closeCtx.Err() != nil {
 				conn.CloseWithError(quic.ApplicationErrorCode(ErrCodeNoError), "")
+				return http.ErrServerClosed
+			}
+
+			// gracefully closed, send GOAWAY frame and wait for requests to complete or grace period to end
+			// new requests will be rejected and shouldn't be sent
+			if s.graceCtx.Err() != nil {
+				b = (&goawayFrame{
+					StreamID: lastID + 4,
+				}).Append(b[:0])
+				// set a deadline to send the GOAWAY frame
+				ctrlStr.SetWriteDeadline(time.Now().Add(goawayTimeout))
+				ctrlStr.Write(b)
+
+				select {
+				// some requests may be still running, wait for them to finish before returning
+				// do nothing let quic to deliver the data the peer
+				case <-handlingDoneChan:
+				case <-s.closeCtx.Done():
+					// close the connection after graceful period
+					conn.CloseWithError(quic.ApplicationErrorCode(ErrCodeNoError), "")
+				}
+				return http.ErrServerClosed
 			}
+
 			var appErr *quic.ApplicationError
 			if errors.As(err, &appErr) && appErr.ErrorCode == quic.ApplicationErrorCode(ErrCodeNoError) {
 				return nil
 			}
 			return fmt.Errorf("accepting stream failed: %w", err)
 		}
-		go s.handleRequest(hconn, str, datagrams, hconn.decoder)
+
+		lastID = str.StreamID()
+		runningRequests.Add(1)
```

Comment:
I don't think we need this tracking here, since it's the client's responsibility to close the connection after receiving GOAWAY and processing in-flight requests.

Reply:
GOAWAY needs to contain the streamID by rfc.

Reply:
Yes, we need `lastID`, but we don't need `runningRequests`.

Reply:
The stream id is used by client to know which request should be retried, because it's not handled by the server.

Reply:
Yes I know, but there's no point in tracking the number of running requests. It's the client's responsibility to close the connection when it's done, as we did in https://github.com/quic-go/quic-go/pull/4669#issuecomment-2403747650.

The server only closes the connection once the context is cancelled, and at that point, we don't care how many requests are active.

Reply:
As above, the context here means the context of the quic.Connection?

Reply:
Server is tracking the number of requests so that closeCtx will be canceled appropriately

Reply:
If you use a `context.Background`, things might block indefinitely. That's expected. Don't do that.

Reply:
That's a normal use of graceful shutdown, which means to wait for all requests to finish indefinitely. stdlib does this too, albeit with polling.

Reply:
And you can do that, but if you're willing to wait indefinitely, you can't complain that you'll wait indefinitely if one of the clients misbehaves...

In practice, you'll *always* apply a timeout.

Reply:
But it should behave correctly when the server is willing to wait. And the server indeed finishes handling the requests.

Reply:
If you're willing to wait forever, the correct behavior is to wait for the client to close the connection.

Reply:
Per stdlib [documents](https://pkg.go.dev/net/http#Server.Shutdown):

>Shutdown gracefully shuts down the server without interrupting any active connections. Shutdown works by first closing all open listeners, then closing all idle connections, and then waiting indefinitely for connections to return to idle and then shut down.

When all requests associated with a quic.Connection are done, it's in the idle state.

Reply:
There is no need to wait for client to close the connection, as newer requests, which shouldn't be sent, will not be handled anymore.

Reply:
I don't see the problem. We already established that the server can't close connections for HTTP/3 without disrupting the client's response processing.

This is exactly what RFC 9114 specifies as well. Let's not violate the RFC.

Reply:
The original problem is that waiting only for closeCtx will lead to deadlock if the grace close context will never be canceled.

The connection can be closed either all requests are done or grace period is over.

Reply:
> I don't see the problem. We already established that the server can't close connections for HTTP/3 without disrupting the client's response processing.
> 
> This is exactly what RFC 9114 specifies as well. Let's not violate the RFC.

If the response can't be guaranteed to succeed, let's just comment out that test. Just let the server close the connection as all requests finish.

Reply:
No, we're not going to disable a test that is correct in order to keep incorrect code around.

I'm getting tired of this. I've said this 5 times now, and you keep ignoring me. I don't know how we should make progress like this.